### ISSUE TITLE
Support for api_private_key

### ISSUE THREAD

**Issue by rbair23** [oracle-terraform-modules/terraform-oci-oke#249]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

In my Terraform scripts I use the "private_key" attribute of the OCI Provider instead of the private_key_path because it is more convenient when used from CI/CD to express the private key as an environment variable rather than having to have some script in the CI/CD workflow write the private key to disk before invoking Terraform. I would like to be able to provide the raw private key to the OKE module as well.

### New or Affected Resource(s)

Module "oke"

**Comment by hyder** [comment]
That's configured on the provider. Since private_key has precedence over private_key_path (https://registry.terraform.io/providers/hashicorp/oci/latest/docs), you can set this accordingly when you configure the provider: https://github.com/oracle-terraform-modules/terraform-oci-oke/blob/master/docs/quickstart.adoc#provisioning-using-this-git-repo

**Comment by rbair23** [comment]
In my own Terraform scripts, that is exactly what I do. I don't see this working with the OKE terraform module however.

```
variable "tenancy_ocid" {}
variable "user_ocid" {}
variable "fingerprint" {}
variable "private_key" {}
variable "private_key_path" {}  # I don't want to use this
variable "private_key_password" {}
variable "region" {}

terraform {
  required_providers {
    oci = {
      source  = "hashicorp/oci"
      version = "~> 4.5.0"
    }
  }
}

# Here I configure my own oci provider, which works if I don't use the OKE module
provider "oci" {
  tenancy_ocid         = var.tenancy_ocid
  user_ocid            = var.user_ocid
  fingerprint          = var.fingerprint
  private_key          = var.private_key
  region               = var.region
  private_key_password = var.private_key_password
}

module "oke" {
  source                  = "oracle-terraform-modules/oke/oci"
  version                 = "2.3.3"

  api_fingerprint         = var.fingerprint
  # This has to be provided or Terraform fails, and I cannot specify api_private_key or api_private_key_password
  api_private_key_path    = var.private_key_path
  compartment_id          = oci_identity_compartment.compartment.id
  tenancy_id              = var.tenancy_ocid
  user_id                 = var.user_ocid

  # ...
}

```

I tried forking the OKE module and adding the missing support, but it still doesn't work. I get an error:

> Error: can not create client, bad configuration: did not find a proper configuration for private key
>
>  on <empty> line 0:
>  (source code not available)

It looks like the problem is not just in the OKE module, but in the base module as well

**Comment by hyder** [comment]
Ah, you're reusing. Yes, I see the api_private_key_path is a string but there's no default value. I think this was because an earlier of the provider only took the path but not the actual key. So, we thought of making it mandatory.

https://github.com/oracle-terraform-modules/terraform-oci-oke/blob/master/variables.tf#L10

We can mark it as optional with the caveat that one or the other must be specified when creating the provider. Would you be willing to do that?

In the meantime, you can just set an empty value for api_private_key_path. Some things might not fully work though like the other submodules (base etc) where we depend on the same mechanims.

**Comment by m1nka** [comment]
I've got the same problem. I tried it in various ways, however auth errors always pop up (e.g. auth for sub modules as you mentioned). Is there a reason why this module does not use the standard OCI naming conventions (`tenancy_ocid`, `user_ocid`, `fingerprint`, ...) including an optional `private_key_password`?

**Comment by hyder** [comment]
They are using standard variables. However, the operator module is using an aliased provider i.e. it's initialising an additional provider using the tenancy's home region instead of just the region where the resources need to be created. This is necessary because a dynamic group is being created and identity operations can only be done in the tenancy's home region.

**Comment by karthicgit** [comment]
@rbair23 @m1nka To be clear on the issue you want to pass private_key instead of the path and a passphrase for the key ? Is it because you want to store this key in Github secrets or something similar and run in a CI/CD tool.

**Comment by hyder** [comment]
https://www.terraform.io/upgrade-guides/0-14.html#sensitive-values-in-plan-output

**Comment by hyder** [comment]
Closed in #372
